lighthouse/multi-cluster-service-selection.md

```diff
@@ -0,0 +1,71 @@
+# Multi-Cluster Service Selection Optimization - EP
+
+## Motivation
+
+While the world is moving towards micro-services application based architecture, more organizations combine multi-cluster and multi-cloud providers for redundancy, performance, management, and other reasons.
+We would like to optimize the total cost of a Kubernetes deployment in means of communication between multi-clusters, given that we could decide between different approaches. The three approaches possible, currently, are optimizing the full service chain of a single app/deployment, multiple apps/deployments, or a smart local optimization per cluster point of view, without building the full service chain.
+In this Enhancement Proposal, we shall introduce questions, ideas, and requirements to optimize the service selection mechanism of the Lighthouse Plugin.
+
+
+### Requirements
+1. Topology:
+    1. Selection Decision should be made locally, i.e per cluster - decentralized solution.
+    2. The solution must be adaptive for topology changes.
+1. Cost Parameters/Metrics:
+    1. Latency
+    2. RTT
+    3. Networking cost
+    4. Network congestion
+    5. CPU/RAM load on each container/pod
+    6. Service scale/autoscale (which actually will affect the load on the CPU and RAM)
+
+
+### Open Questions
+1. Topology:
+    1. What topology would we like to optimize?
+        1. Local optimization per cluster?
+        2. Per App optimization?
+        3. Multi App optimization?
+    2. How often do deployments affects changes in the topology?
+    3. How adaptive would we like to be for changes in the topology?
+    4. Would we like to build the full application graph for better optimization?
+        1.  How much are we willing to pay for the data sharing through the broker system?
+1. Cost Parameters/Metrices:
+    1.  How often would we like to collection the parameters?
+    2.  Would we like to make the collection configurable?
+1.  Optimization:
+    1. Would we like to take the processing time of each service in account?
+
+
+### Design Details
+As we would like to comply with the Kubernetes KEP for the Multi-Cluster API, and the current implementation, we would like to offer a simple lightweight solution for selecting a service from a given pool.
+
+#### Priority Queue
+
+It seems that a data structure of a priority queue could be a simple elegant solution for selecting the best service per request.

```

Comment:
It's not really clear how this would work though, when do you remove and add items to the queue? How do you make sure that you're not bombarding a single service while starving the others? What about connection stickiness/tracking (eg TCP needs to reach the same service each time)?